### PR TITLE
Built-in features in class ROUTINE and descendants

### PR DESCRIPTION
**Pull request by ebezault**
Added built-in features in Agent classes, so that the same code can be compiled by ISE and Gobo even if they have different compilation techniques.

### CODE REVIEW

File: Src/library/base/elks/kernel/function.e
Comment (ebezault) [inline Src/library/base/elks/kernel/function.e:27]: 'set_rout_disp_final' is used in finalized mode to create agent objects from the runtime.

File: Src/library/base/elks/kernel/function.e
Comment (ebezault) [inline Src/library/base/elks/kernel/function.e:45]: The call to 'fast_item' is too specific to the ISE compiler, hence the addition of these two built-in features. Furthermore, the fact that 'item' is built-in allows the Gobo compiler to optimize the code when the argument is a manifest tuple.

File: Src/library/base/elks/kernel/routine.e
Comment (ebezault) [inline Src/library/base/elks/kernel/routine.e:179]: Believe it or not, this new implementation is on average faster than the old one. And it avoids the call to the ISE's specific external routine 'eif_gen_typecode_str'.
My benchmarks, calling 10,000,000 times 'valid_operands'.
Argument containing only references: old 2.7 seconds, new 1.5 seconds
Argument containing more basic types than references: old: 2.2 seconds, new: 1.4 seconds
Now, resetting 'open_types' to Void before each call:
Argument containing only references: old 6.3 seconds, new 5.2 seconds
Argument containing more basic types than references: old: 3.7 seconds, new: 5.3 seconds
This last one is the only case where the old implementation is faster.

File: Src/library/base/elks/kernel/routine.e
Comment (ebezault) [inline Src/library/base/elks/kernel/routine.e:197]: There is no assertion which states that:
   is_target_closed implies closed_count >  0
   not is_target_closed implies open_count >  0
Without these assertions, the new implementation above is more accurate.

File: Src/library/base/elks/kernel/routine.e
Comment (ebezault) [inline Src/library/base/elks/kernel/routine.e:417]: I tried to remove this C function from the runtime, but it is still used in:
eweasel/tests/term036/routine.e
eweasel/tests/valid068/routine.e
historic/library/base_original/kernel/classic/routine.e
Src/experimental/library/base/elks/kernel/routine.e

File: Src/library/base/elks/kernel/routine.e
Comment (ebezault) [inline Src/library/base/elks/kernel/routine.e:355]: It would be nice if all attributes could have a header comment, explaining what they are used for and whether they are used in workbench or finalized mode or both.

**Comment by ebezault** [comment]
Any comment on that?

**Comment by ebezault** [comment]
Is there a way to contribute to EiffelStudio, its compiler and libraries, without having to wait for months to get an answer?
Can we gain some Eiffel skill/trust points somehow, and have write access to the repository above some point threshold?

**Comment by Bertrand-Meyer** [comment]
Very good point. We must put a process in place. Anyone interested in starting to contribute in the meantime (while the general process is being set up) is welcome to contact me directly so that we can start with an informal procedure.

File: Src/library/base/elks/kernel/routine.e
Comment (manuseiffel) [reply Src/library/base/elks/kernel/routine.e:417]: We will update those classes and remove this C routine from the runtime.

**Review by manuseiffel — APPROVED** [review]

**Comment by ebezault** [comment]
Committed in svn#100440.

**Pull request closed without merging**
